### DISCUSSION TITLE
Add SGM Low RAM and SGM High RAM to memory map for Colecovision

### DISCUSSION THREAD

**Discussion by ladynadiad** [RetroAchievements/RAIntegration#1104] [Ideas]
As of now on Colecovision only 1024 bytes of RAM in Main RAM are accessible via the memory inspector.  However, I found out earlier that Bizhawk does have the capability of running games that use the Super Game Module which BlueMSX seems to lack.  Because of this, it would be nice if the ranges unique to that could be added to the map so it would be viable to make sets for these games.  ColecoHawk definitely is able to expose it and I was able to find values just fine via its RAM search.  The games seem to hash just fine and run great in ColecoHawk.

**Comment by Jamiras** [comment]
[SGM](https://www.colecovision.dk/sem.htm?refreshed) is an aftermarket product released in 2012.

From looking at https://forums.atariage.com/topic/203851-super-game-module-technical-forum/, it appears that the 32KB of additional memory is exposed from $0000-$7FFF, preventing access to the real memory at $6000-$63FF, and the BIOS at $0000.

Since it's hiding the existing RAM, we'd probably have to expose it using virtual addresses (like the paged GBC banks), which requires additional work from cores/emulators.

**Comment by Jamiras** [comment]
> I found out earlier that Bizhawk does have the capability of running games that use the Super Game Module which BlueMSX seems to lack.

It seems like blueMSX defaults to MSX mode with 512KB of memory exposed when loading these games. The game doesn't run, but it might be an awkward way to false-trigger achievements in these games. We'll probably need some way to detect and prevent that behavior similar to how we [protect against UniBIOS for Neo Geo](https://docs.retroachievements.org/developer-docs/console-specific-tips.html#mandatory-unibios-protection).

**Comment by Jamiras** [comment]
gearcoleco does allow the games to run, but only exposes the [BIOS memory and system memory](https://github.com/drhelius/Gearcoleco/blob/ccaa62fac4762b121627a303e4aefcde9a26500d/platforms/libretro/libretro.cpp#L635-L650).

We've also [marked it as unsupported](https://docs.retroachievements.org/general/emulator-support-and-issues.html#colecovision). As best as I can tell from searching discord, some sets just don't work with it at all, and other times it has some input problems where buttons appear to get stuck, making the games harder than they should be.

**Comment by ladynadiad** [comment]
Definitely agree that protection will be needed for unsupported cores, though truly I think that should get a mention in the docs for in general with any systems that have BlueMSX as a supported core since problems like that can come up with MSX and SG-1000 too.  Wrong machine setting can cause all sorts of havoc with sets and in some cases the games DO work.
